File: src/clock.py
import ticks_minimal as ticks
from settings import settings

class Clock:
    """MIDI clock: tracks ticks, BPM, and note durations."""

    MILLISECONDS_TO_SECONDS = 1000.0
    TICKS_PER_QUARTER_NOTE = 24
    TICKS_PER_WHOLE_NOTE = TICKS_PER_QUARTER_NOTE * 4

    def __init__(self):
        self.midi_ticks_elapsed = 0                   
        self.last_tick_time = ticks.ticks_ms()
        self.bpm_current = 120.0
        self.set_bpm(self.bpm_current)
        self.is_playing = False
        self.last_whole_note_time = 0 
        self.new_tick = False
        self.pending_bpm = None  # BPM waiting for confirmation (filters glitches) 

    def reset_midi_tick_count(self):
        self.midi_ticks_elapsed = 0

    def set_bpm(self, bpm):
        quarternote_duration = 60 / bpm
        self.bpm_current = bpm
        self.seconds_per_tick = 60 / (self.bpm_current * self.TICKS_PER_QUARTER_NOTE)
        self.quarternote_duration = quarternote_duration
        self.halfnote_duration = quarternote_duration * 2
        self.wholenote_duration = quarternote_duration * 4
        self.eighthnote_duration = quarternote_duration / 2
        self.sixteenthnote_duration = quarternote_duration / 4
        
    def reset_new_tick_flag(self):
        self.new_tick = False

    def update_clock(self):
        """Update on new MIDI clock tick. Calculates BPM from tick intervals."""
        self.new_tick = True
        self.midi_ticks_elapsed += 1

        timenow = ticks.ticks_ms()
        self.last_tick_time = timenow

        if self.midi_ticks_elapsed % self.TICKS_PER_WHOLE_NOTE == 0:
            whole_note_time = ticks.ticks_diff(timenow, self.last_whole_note_time) / self.MILLISECONDS_TO_SECONDS
            self.last_whole_note_time = timenow
            
            if whole_note_time > 0:
                new_bpm = round(60 * 4 / whole_note_time) # Divide by 4 to get the BPM from whole note time
            else:
                new_bpm = 0

            if new_bpm != self.bpm_current and new_bpm > 0:
                # Ignore ±1 BPM fluctuations (measurement noise from timing jitter)
                if abs(new_bpm - self.bpm_current) <= 1:
                    self.pending_bpm = None
                elif new_bpm == self.pending_bpm:
                    if settings.debug:
                        print(f"[DEBUG] BPM changed: {self.bpm_current} -> {new_bpm}")
                    self.set_bpm(new_bpm)
                    self.pending_bpm = None
                else:
                    # First time seeing this value, wait for confirmation
                    self.pending_bpm = new_bpm
            else:
                # Current BPM is stable, clear any pending
                self.pending_bpm = None

    def seconds_to_ticks(self, seconds, bpm=None):
        """Convert seconds to MIDI ticks. Uses current BPM if not specified."""
        if bpm:
            seconds_per_tick = 60 / (bpm * self.TICKS_PER_QUARTER_NOTE)
        else:
            seconds_per_tick = self.seconds_per_tick
        return int(round(seconds / seconds_per_tick))
    
    def ticks_to_seconds(self, ticks, bpm=None):
        """Convert MIDI ticks to seconds. Uses current BPM if not specified."""
        if bpm:
            seconds_per_tick = 60 / (bpm * self.TICKS_PER_QUARTER_NOTE)
        else:
            seconds_per_tick = self.seconds_per_tick
        return ticks * seconds_per_tick

    def get_note_duration_seconds(self, note_type):
        """Get duration in seconds for note type (e.g., 'quarter', '1/4', 'whole')."""
        note_times_seconds = {
            "whole": self.wholenote_duration,
            "1": self.wholenote_duration,
            "half": self.halfnote_duration,
            "1/2": self.halfnote_duration,
            "quarter": self.quarternote_duration,
            "1/4": self.quarternote_duration,
            "eighth": self.eighthnote_duration,
            "1/8": self.eighthnote_duration,
            "sixteenth": self.sixteenthnote_duration,
            "1/16": self.sixteenthnote_duration,
            "thirtysecond": self.sixteenthnote_duration / 2,
            "1/32": self.sixteenthnote_duration / 2,
            "sixtyfourth": self.sixteenthnote_duration / 4,
            "1/64": self.sixteenthnote_duration / 4,
        }

        return note_times_seconds.get(note_type, self.quarternote_duration)

    def start_clock(self):
        self.is_playing = True
        self.reset_midi_tick_count()
        self.last_whole_note_time = ticks.ticks_ms()  # Fresh timing reference
        self.pending_bpm = None  # Clear any pending BPM
        self.new_tick = True

    def continue_clock(self):
        """Resume clock without resetting tick count (MIDI Continue)."""
        self.is_playing = True
        self.last_whole_note_time = ticks.ticks_ms()  # Fresh timing reference for BPM
        self.pending_bpm = None  # Clear any pending BPM to avoid glitches
        # Don't reset midi_ticks_elapsed - Continue resumes from current position
    
    def stop_clock(self):
        self.is_playing = False
        # Don't reset midi_ticks_elapsed - needed for recording finalization
        # when Stop is received before toggle_record_state() calculates total_midi_ticks

    def get_playstate(self):
        return self.is_playing

clock = Clock()

File: src/clock.cpp
#include "clock.h"
#include "ticks.h"
#include "settings.h"
#include "serial_config.h" // cdc_log (Q1 bounded prints)

Clock clock_;

Clock::Clock() {
    last_tick_time = 0; // millis() not valid in static-init; harmless (Python used boot time)
    set_bpm(bpm_current);
}

void Clock::set_bpm(float bpm) {
    float quarter = 60.0f / bpm;
    bpm_current = bpm;
    seconds_per_tick = 60.0f / (bpm_current * TICKS_PER_QUARTER_NOTE);
    quarternote_duration = quarter;
    halfnote_duration = quarter * 2;
    wholenote_duration = quarter * 4;
    eighthnote_duration = quarter / 2;
    sixteenthnote_duration = quarter / 4;
}

void Clock::update_clock() {
    uint32_t timenow = ticks::ticks_ms();
    last_tick_time = timenow;

    if (first_tick_pending) {
        // Downbeat tick: counter stays 0 so midi_ticks_elapsed == song tick from here on
        // (see clock.h). BPM window restarts here so the first beat measurement
        // spans exactly 24 tick intervals instead of 25. new_tick stays clear too: a loop
        // resumed at Start is already AT position 0 == this downbeat, so advancing playback
        // here made the entire first pass play one tick early (hw-observed: ~19 ms @ 130,
        // self-corrected only at the first wrap re-snap).
        first_tick_pending = false;
        bpm_window_start_time = timenow;
        bpm_window_ticks = 0;
        return;
    }
    new_tick = true;
    midi_ticks_elapsed++;

    // Quarter-note measurement window (was whole-note in the Python port). The 96-tick
    // window + double-confirm was sized for CircuitPython main-loop stalls; with ms
    // timestamps only the window endpoints matter, so one beat (±2-4 ms endpoint jitter
    // over ~500 ms @ 120) resolves integer BPM within the ±1 deadband. Latch worst case
    // drops from ~3 whole notes to 3 beats. Confirmation + deadband kept: consecutive
    // windows share the boundary timestamp, so a jitter spike skews them in opposite
    // directions and can never confirm. Window ticks are counted explicitly (not
    // midi_ticks_elapsed % 24) so an SPP+Continue resume mid-beat can't produce a
    // short first window and a bogus over-read.
    if (++bpm_window_ticks >= TICKS_PER_QUARTER_NOTE) {
        bpm_window_ticks = 0;
        float beat_time = ticks::ticks_diff(timenow, bpm_window_start_time) / MILLISECONDS_TO_SECONDS;
        bpm_window_start_time = timenow;

        int new_bpm = 0;
        if (beat_time > 0) {
            new_bpm = (int)roundf(60.0f / beat_time);
        }

        if (new_bpm != (int)bpm_current && new_bpm > 0) {
            // Ignore ±1 BPM fluctuations (measurement noise from timing jitter)
            if (abs(new_bpm - (int)bpm_current) <= 1) {
                pending_bpm = -1;
            } else if (new_bpm == pending_bpm) {
                set_bpm((float)new_bpm);
                pending_bpm = -1;
            } else {
                // First time seeing this value, wait for confirmation
                pending_bpm = new_bpm;
            }
        } else {
            // Current BPM is stable, clear any pending
            pending_bpm = -1;
        }
    }
}

int32_t Clock::seconds_to_ticks(float seconds, float bpm) const {
    float spt = (bpm > 0) ? 60.0f / (bpm * TICKS_PER_QUARTER_NOTE) : seconds_per_tick;
    return (int32_t)roundf(seconds / spt);
}

float Clock::ticks_to_seconds(int32_t t, float bpm) const {
    float spt = (bpm > 0) ? 60.0f / (bpm * TICKS_PER_QUARTER_NOTE) : seconds_per_tick;
    return t * spt;
}

float Clock::get_note_duration_seconds(const char *note_type) const {
    if (!strcmp(note_type, "whole") || !strcmp(note_type, "1")) return wholenote_duration;
    if (!strcmp(note_type, "half") || !strcmp(note_type, "1/2")) return halfnote_duration;
    if (!strcmp(note_type, "quarter") || !strcmp(note_type, "1/4")) return quarternote_duration;
    if (!strcmp(note_type, "eighth") || !strcmp(note_type, "1/8")) return eighthnote_duration;
    if (!strcmp(note_type, "sixteenth") || !strcmp(note_type, "1/16")) return sixteenthnote_duration;
    if (!strcmp(note_type, "thirtysecond") || !strcmp(note_type, "1/32")) return sixteenthnote_duration / 2;
    if (!strcmp(note_type, "sixtyfourth") || !strcmp(note_type, "1/64")) return sixteenthnote_duration / 4;
    return quarternote_duration;
}

void Clock::start_clock() {
    is_playing = true;
    reset_midi_tick_count();
    first_tick_pending = true; // next 0xF8 is the downbeat — don't count it (see clock.h)
    bpm_window_start_time = ticks::ticks_ms(); // fresh timing reference
    bpm_window_ticks = 0;
    pending_bpm = -1;
    // Deliberately no new_tick here (removed with the downbeat fix): this synthetic tick
    // pushed resumed loops to position 1 while the song was still at tick 0 — the other
    // half of the first-pass-early skew. Tick-0 events don't need it; get_new_events
    // fires `tick <= current` even without an increment.
}

void Clock::continue_clock() {
    is_playing = true;
    // The clock bundled with Continue IS the resume-point tick (see first_tick_pending in
    // clock.h) — the counter already holds that position (via SPP, or the stop-point count),
    // so counting it would put the grid one tick early.
    first_tick_pending = true;
    bpm_window_start_time = ticks::ticks_ms(); // fresh timing reference for BPM
    bpm_window_ticks = 0;
    pending_bpm = -1;
    // Don't reset midi_ticks_elapsed — SPP (sent by Live right before Continue) pins it;
    // without SPP it resumes from the stop-point count
}

void Clock::stop_clock() {
    is_playing = false;
    // Don't reset midi_ticks_elapsed — needed for recording finalization
    // when Stop arrives before toggle_record_state() computes total_midi_ticks
}
